pkg/docker/config/config.go

```diff
@@ -0,0 +1,281 @@
+// either auth.json file or .docker/config.json
+func GetAuthentication(registry string) (string, string, error) {
+	registry = DockerHostNameToRegistry(registry)
+	// Check auth.json file first
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	username, password, err := findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if username != "" && password != "" {
+		return username, password, nil
+	}
+
+	// check ~/.docker/config.json as secondary option
+	path = filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName)
+	username, password, err = findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	return username, password, nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) error {
+	registry = DockerHostNameToRegistry(registry)
+	path, err := getPathToAuth()
+	if err != nil {
+		return err
+	}
+	return modifyJSON(registry, "", "", path, deleteAuth)
+}
+
+// DockerHostNameToRegistry converts the docker hostname to the docker registry
+func DockerHostNameToRegistry(registry string) string {
+	if registry == dockerHostname {
+		return dockerRegistry
+	}
+	return registry
+}
+
+// getPath gets the path of the auth.json file since the user can set their own path
+// by setting CRIO_AUTH_PATH
+// If CRIO_AUTH_PATH is not set, the default path is ~/.config/containers/auth.json
+func getPathToAuth() (string, error) {
+	if os.Getenv("CRIO_AUTH_PATH") != "" {
+		return filepath.Join(os.Getenv("CRIO_AUTH_PATH"), authCfgFileName), nil
+	}
+	dir := filepath.Join(homedir.Get(), authCfg)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err = os.Mkdir(dir, 0700); err != nil {
+			return "", errors.Wrapf(err, "error creating directory %q", dir)
+		}
+	}
+	return filepath.Join(dir, authCfgFileName), nil
+}
+
+// readJSONFile unmarshals the authentications stored in the auth.json file and returns it
+// or returns an empty dockerConfigFile data structure if auth.json is empty
+func readJSONFile(path string) (dockerConfigFile, error) {
+	var auths dockerConfigFile
+
+	raw, err := ioutil.ReadFile(path)
+	if os.IsNotExist(err) {
+		auths.AuthConfigs = map[string]dockerAuthConfig{}
+		return auths, nil
+	}
+
+	if err = json.Unmarshal(raw, &auths); err != nil {
+		return dockerConfigFile{}, errors.Wrapf(err, "error unmarshaling JSON at %q", path)
+	}
+
+	return auths, nil
+}
+
+// addAuthToJSON either adds a new registry url with its encoded user:pass
+// or modifies an existing registry url
+func modifyJSON(registry, username, password, path, action string) error {
+	auths, err := readJSONFile(path)
+	if err != nil {
+		return errors.Wrapf(err, "error reading JSON file %q", path)
+	}
+
+	switch action {
+	case addAuth:
+		if len(auths.CredHelpers) > 0 {
+			if err = setAuthToCredHelper(registry, username, password); err != nil {
+				return err
+			}
+			break
```

Comment:
Just a `return setAuthToCredHelper(…)`?  We don’t need to write the unmodified config file back.

(Same in the delete case).

Reply:
changed the code around a bit. fixed